README.md
dlz-mysql-dynamic
==================

A Bind 9 Dynamically Loadable MySQL Driver

Summary
-------

This is an attempt to port the original Bind 9 DLZ dlz_mysql_driver.c as
found in the Bind 9 source tree into the new DLZ dlopen driver API.
The goals of this project are as follows:

* Provide DLZ facilities to OEM-supported Bind distributions
* Support both v1 (Bind 9.8) and v2 (Bind 9.9) of the dlopen() DLZ API

Requirements
------------

You will need the following:
 * Bind 9.8 or higher with the DLZ dlopen driver enabled
 * MySQL client libraries and source headers
 * A C compiler

Installation
------------

With the above requirements satisfied perform the following steps:

1. Ensure the symlink for dlz_minimal.h points at the correct header
   file matching your Bind version
2. Run: make
3. Run: sudo make install # this will install dlz_mysql_dynamic.so
   into /usr/lib/bind9/
4. Add a DLZ statement similar to the example shown in
   example/dlz.conf into your Bind configuration
5. Create a MySQL database and schema to support your data XXX UPDATE ME XXX   
6. If you're running an AppArmor enabled Bind, consider adding content
   included within example/apparmor.d-local-usr.sbin.named within
   /etc/apparmor.d/local/usr.sbin.named
7. Use the included testing/mysql-populate.pl script to provide some
   data for initial testing XXX UPDATE ME XXX

Usage
-----

Example usage is as follows:

```
dlz "mysql_dynamic" {
        database "dlopen /usr/lib/bind9/dlz_mysql_dynamic.so 
        {host=localhost dbname=dns_data ssl=true}
        {select zone from dns_records where zone = '%zone%'}
        {select ttl, type, mx_priority, case when lower(type)='txt' then concat('\"', data, '\"')
           when lower(type) = 'soa' then concat_ws(' ', data, resp_person, serial, refresh, retry, expire, minimum)
           else data end from dns_records where zone = '%zone%' and host = '%record%'}
        {select ttl, type, host, mx_priority, data, resp_person, serial, refresh, retry, expire, minimum 
                from dns_records where zone = '%zone%' and not (type = 'SOA' or type = 'NS')}
        {select ttl, type, host, mx_priority, case when lower(type)='txt' then
           concat('\"', data, '\"') else data end, resp_person, serial, refresh, retry, expire,
           minimum from dns_records where zone = '%zone%'}
        {select zone from xfr_table where zone = '%zone%' and client = '%client%'}
        {update data_count set count = count + 1 where zone ='%zone%'}";
};
```

The arguments for the "database" line above are as follows:

1. dlopen - Use the dlopen DLZ driver to dynamically load our compiled
   driver
2. The full path to your built dlz_mysql_dynamic.so
3. Connection options for this MySQL connection, supported options:
   - dbname
   - port
   - host
   - user
   - pass
   - socket
   - compress
   - ssl
   - space
4. Query to ascertain whether we host this zone - used by findzone() - Required
5. Query to find individual record(s) - used by lookup() - Required
6. Query to satisfy authority section lookups - used by authority() - Optional
7. Query to find all records for a zone - used by allnodes() - Optional
8. Query to establish whether a client is authorised to perform - Optional
   zone-transfer, requires prior query to be specified to work. - Optional
9. Query to update a counter of queries performed against a given zone - Optional

A copy of the above Bind configuration is included within
example/dlz.conf.

Author
------

The person responsible for this is:

 Mark Goldfinch <g@g.org.nz>

Original license conditions from both dlz_mysql_driver.c and
dlz_example.c are maintained in dlz_mysql_dynamic.c.


debian/changelog
libmysql-dynamic-bind9 (0.1) precise; urgency=low

  * Initial release.

 -- Mark Goldfinch <mark.goldfinch@g.org.nz>  Sat, 25 Aug 2012 20:16:24 +1200


include/dlz/sdlz_helper.h
/*
 * Copyright (C) 2002 Stichting NLnet, Netherlands, stichting@nlnet.nl.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all
 * copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND STICHTING NLNET
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
 * STICHTING NLNET BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
 * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
 * OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE
 * USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * The development of Dynamically Loadable Zones (DLZ) for Bind 9 was
 * conceived and contributed by Rob Butler.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all
 * copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND ROB BUTLER
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
 * ROB BUTLER BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
 * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
 * OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE
 * USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#ifndef SDLZHELPER_H
#define SDLZHELPER_H

/*
 * Types
 */
#define SDLZH_REQUIRE_CLIENT	0x01
#define SDLZH_REQUIRE_QUERY	0x02
#define SDLZH_REQUIRE_RECORD	0x04
#define SDLZH_REQUIRE_ZONE	0x08

typedef struct query_segment query_segment_t;
typedef ISC_LIST(query_segment_t) query_list_t;
typedef struct mysql_instance mysql_instance_t;
typedef ISC_LIST(mysql_instance_t) db_list_t;
typedef struct driverinstance driverinstance_t;

/*%
 * a query segment is all the text between our special tokens
 * special tokens are %zone%, %record%, %client%
 */
struct query_segment {
	void				*sql;
	unsigned int			strlen;
	isc_boolean_t			direct;
	ISC_LINK(query_segment_t)	link;
};

struct mysql_instance {
  void			  *dbconn;
  query_list_t		  *allnodes_q;
  query_list_t		  *allowxfr_q;
  query_list_t		  *authority_q;
  query_list_t		  *findzone_q;
  query_list_t		  *lookup_q;
  query_list_t		  *countzone_q;
  char			  *query_buf;
  char			  *zone;
  char			  *record;
  char			  *client;

  /* Helper functions from the dlz_dlopen driver */
  log_t                   *log;
  dns_sdlz_putrr_t        *putrr;
  dns_sdlz_putnamedrr_t   *putnamedrr;
  dns_dlz_writeablezone_t *writeable_zone;
};

/*
 * Method declarations
 */

/* see the code in sdlz_helper.c for more information on these methods */

char *
sdlzh_build_querystring(isc_mem_t *mctx, query_list_t *querylist);

isc_result_t
sdlzh_build_sqldbinstance(isc_mem_t *mctx, const char *allnodes_str,
			  const char *allowxfr_str, const char *authority_str,
			  const char *findzone_str, const char *lookup_str,
			  const char *countzone_str, mysql_instance_t **dbi);

void
sdlzh_destroy_sqldbinstance(mysql_instance_t *dbi);

char *
sdlzh_get_parameter_value(isc_mem_t *mctx, const char *input, const char* key);

/* Compatability with existing DLZ drivers */
#define	build_querystring	sdlzh_build_querystring
#define	build_sqldbinstance	sdlzh_build_sqldbinstance
#define	destroy_sqldbinstance	sdlzh_destroy_sqldbinstance

#define	getParameterValue(x,y)  sdlzh_get_parameter_value(ns_g_mctx, (x), (y))

#endif /* SDLZHELPER_H */


include/dlz/dlz_ldopen_v1.h
/*
 * Copyright (C) 2010 Andrew Tridgell
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all
 * copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
 * THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
 * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
 * OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE
 * USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
  This header provides a minimal set of defines and typedefs needed
  for building an external DLZ module for bind9. When creating a new
  external DLZ driver, please copy this header into your own source
  tree.
 */
typedef unsigned int isc_result_t;
typedef int isc_boolean_t;
typedef uint32_t dns_ttl_t;

#define DLZ_DLOPEN_VERSION 1

/* return this in flags to dlz_version() if thread safe */
#define DNS_SDLZFLAG_THREADSAFE		0x00000001U

/* result codes */
#define ISC_R_SUCCESS			0
#define ISC_R_NOMEMORY			1
#define ISC_R_NOTFOUND			23
#define ISC_R_FAILURE			25

/* log levels */
#define ISC_LOG_DEBUG(level)    (level)
#define ISC_LOG_INFO		(-1)
#define ISC_LOG_NOTICE		(-2)
#define ISC_LOG_WARNING 	(-3)
#define ISC_LOG_ERROR		(-4)
#define ISC_LOG_CRITICAL	(-5)

/* other useful definitions */
#define UNUSED(x) (void)(x)

/* some opaque structures */
typedef void *dns_sdlzlookup_t;
typedef void *dns_sdlzallnodes_t;
typedef void *dns_view_t;

/*
 * Method definitions for callbacks provided by the dlopen driver
 */
typedef void log_t(int level, const char *fmt, ...);

typedef isc_result_t dns_sdlz_putrr_t(dns_sdlzlookup_t *lookup,
                                      const char *type,
                                      dns_ttl_t ttl,
                                      const char *data);

typedef isc_result_t dns_sdlz_putnamedrr_t(dns_sdlzallnodes_t *allnodes,
                                           const char *name,
                                           const char *type,
                                           dns_ttl_t ttl,
                                           const char *data);

typedef isc_result_t dns_dlz_writeablezone_t(dns_view_t *view,
                                             const char *zone_name);


/*
 * prototypes for the functions you can include in your driver
 */


/*
 * dlz_version() is required for all DLZ external drivers. It should
 * return DLZ_DLOPEN_VERSION
 */
int dlz_version(unsigned int *flags);

/*
 * dlz_create() is required for all DLZ external drivers.
 */
isc_result_t dlz_create(const char *dlzname, unsigned int argc, char *argv[], void **dbdata, ...);

/*
 * dlz_destroy() is optional, and will be called when the driver is
 * unloaded if supplied
 */
void dlz_destroy(void *dbdata);

/*
  dlz_findzonedb is required for all DLZ external drivers
 */
isc_result_t dlz_findzonedb(void *dbdata, const char *name);

/*
  dlz_lookup is required for all DLZ external drivers
 */
isc_result_t dlz_lookup(const char *zone, const char *name,
			void *dbdata, dns_sdlzlookup_t *lookup);

/*
  dlz_allowzonexfr() is optional, and should be supplied if you want
  to support zone transfers
 */
isc_result_t dlz_allowzonexfr(void *dbdata, const char *name, const char *client);


/*
  dlz_allnodes() is optional, but must be supplied if supply a
  dlz_allowzonexfr() function
 */
isc_result_t dlz_allnodes(const char *zone, void *dbdata, dns_sdlzallnodes_t *allnodes);

/*
  dlz_newversion() is optional. It should be supplied if you want to
  support dynamic updates.
 */
isc_result_t dlz_newversion(const char *zone, void *dbdata, void **versionp);

/* 
   dlz_closeversion() is optional, but must be supplied if you supply
   a dlz_newversion() function
 */
void dlz_closeversion(const char *zone, isc_boolean_t commit, void *dbdata, void **versionp);

/*
  dlz_configure() is optional, but must be supplied if you want to
  support dynamic updates
 */
isc_result_t dlz_configure(dns_view_t *view, void *dbdata);

/*
  dlz_ssumatch() is optional, but must be supplied if you want to
  support dynamic updates
 */
isc_boolean_t dlz_ssumatch(const char *signer, const char *name, const char *tcpaddr,
			   const char *type, const char *key, uint32_t keydatalen, uint8_t *keydata,
			   void *dbdata);

/*
  dlz_addrdataset() is optional, but must be supplied if you want to
  support dynamic updates
 */
isc_result_t dlz_addrdataset(const char *name, const char *rdatastr, void *dbdata, void *version);

/*
  dlz_subrdataset() is optional, but must be supplied if you want to
  support dynamic updates
 */
isc_result_t dlz_subrdataset(const char *name, const char *rdatastr, void *dbdata, void *version);

/*
  dlz_delrdataset() is optional, but must be supplied if you want to
  support dynamic updates
 */
isc_result_t dlz_delrdataset(const char *name, const char *type, void *dbdata, void *version);


include/dlz/dlz_ldopen_v2.h
/*
 * Copyright (C) 2011  Internet Systems Consortium, Inc. ("ISC")
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all
 * copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
 * THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
 * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
 * OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE
 * USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * This header provides a minimal set of defines and typedefs needed
 * for building an external DLZ module for bind9. When creating a new
 * external DLZ driver, please copy this header into your own source
 * tree.
 */

#include <sys/types.h>
#include <sys/socket.h>
#ifdef ISC_PLATFORM_HAVESYSUNH
#include <sys/un.h>
#endif
#include <net/if.h>
#include <netinet/in.h>
#include <arpa/inet.h>

typedef unsigned int isc_result_t;
typedef int isc_boolean_t;
typedef uint32_t dns_ttl_t;

#define DLZ_DLOPEN_VERSION 2

/* return this in flags to dlz_version() if thread safe */
#define DNS_SDLZFLAG_THREADSAFE		0x00000001U

/* result codes */
#define ISC_R_SUCCESS			0
#define ISC_R_NOMEMORY			1
#define ISC_R_NOPERM			6
#define ISC_R_NOTFOUND			23
#define ISC_R_FAILURE			25
#define ISC_R_NOMORE			29

/* boolean values */
#define ISC_TRUE 1
#define ISC_FALSE 0

/* log levels */
#define ISC_LOG_DEBUG(level)    (level)
#define ISC_LOG_INFO		(-1)
#define ISC_LOG_NOTICE		(-2)
#define ISC_LOG_WARNING 	(-3)
#define ISC_LOG_ERROR		(-4)
#define ISC_LOG_CRITICAL	(-5)

/* other useful definitions */
#define UNUSED(x) (void)(x)

/* opaque structures */
typedef void *dns_sdlzlookup_t;
typedef void *dns_sdlzallnodes_t;
typedef void *dns_view_t;

/*
 * Method and type definitions needed for retrieval of client info
 * from the caller.
 */
typedef struct isc_sockaddr {
        union {
		struct sockaddr         sa;
		struct sockaddr_in      sin;
		struct sockaddr_in6     sin6;
#ifdef ISC_PLATFORM_HAVESYSUNH
		struct sockaddr_un      sunix;
#endif
	}                               type;
	unsigned int                    length;
	void *				link;
} isc_sockaddr_t;

#define DNS_CLIENTINFO_VERSION 1
typedef struct dns_clientinfo {
	uint16_t version;
	void *data;
} dns_clientinfo_t;

typedef isc_result_t (*dns_clientinfo_sourceip_t)(dns_clientinfo_t *client,
						  isc_sockaddr_t **addrp);

#define DNS_CLIENTINFOMETHODS_VERSION 1
#define DNS_CLIENTINFOMETHODS_AGE 0

typedef struct dns_clientinfomethods {
	uint16_t version;
	uint16_t age;
	dns_clientinfo_sourceip_t sourceip;
} dns_clientinfomethods_t;

/*
 * Method definitions for callbacks provided by the dlopen driver
 */
typedef void log_t(int level, const char *fmt, ...);

typedef isc_result_t dns_sdlz_putrr_t(dns_sdlzlookup_t *lookup,
				      const char *type,
				      dns_ttl_t ttl,
				      const char *data);

typedef isc_result_t dns_sdlz_putnamedrr_t(dns_sdlzallnodes_t *allnodes,
					   const char *name,
					   const char *type,
					   dns_ttl_t ttl,
					   const char *data);

typedef isc_result_t dns_dlz_writeablezone_t(dns_view_t *view,
					     const char *zone_name);


/*
 * prototypes for the functions you can include in your module
 */


/*
 * dlz_version() is required for all DLZ external drivers. It should
 * return DLZ_DLOPEN_VERSION.  'flags' is updated to indicate capabilities
 * of the module.  In particular, if the module is thread-safe then it
 * sets 'flags' to include DNS_SDLZFLAG_THREADSAFE.  Other capability
 * flags may be added in the future.
 */
int
dlz_version(unsigned int *flags);

/*
 * dlz_create() is required for all DLZ external drivers.
 */
isc_result_t
dlz_create(const char *dlzname, unsigned int argc, char *argv[],
           void **dbdata, ...);

/*
 * dlz_destroy() is optional, and will be called when the driver is
 * unloaded if supplied
 */
void
dlz_destroy(void *dbdata);

/*
 * dlz_findzonedb is required for all DLZ external drivers
 */
isc_result_t
dlz_findzonedb(void *dbdata, const char *name);

/*
 * dlz_lookup is required for all DLZ external drivers
 */
isc_result_t
dlz_lookup(const char *zone, const char *name, void *dbdata,
	   dns_sdlzlookup_t *lookup,
	   dns_clientinfomethods_t *methods,
	   dns_clientinfo_t *clientinfo);

/*
 * dlz_allowzonexfr() is optional, and should be supplied if you want to
 * support zone transfers
 */
isc_result_t
dlz_allowzonexfr(void *dbdata, const char *name, const char *client);


/*
 * dlz_allnodes() is optional, but must be supplied if supply a
 * dlz_allowzonexfr() function
 */
isc_result_t
dlz_allnodes(const char *zone, void *dbdata, dns_sdlzallnodes_t *allnodes);

/*
 * dlz_newversion() is optional. It should be supplied if you want to
 * support dynamic updates.
 */
isc_result_t
dlz_newversion(const char *zone, void *dbdata, void **versionp);

/* 
 * dlz_closeversion() is optional, but must be supplied if you supply a
 * dlz_newversion() function
 */
void
dlz_closeversion(const char *zone, isc_boolean_t commit, void *dbdata,
                 void **versionp);

/*
 * dlz_configure() is optional, but must be supplied if you want to support
 * dynamic updates
 */
isc_result_t
dlz_configure(dns_view_t *view, void *dbdata);

/*
 * dlz_ssumatch() is optional, but must be supplied if you want to support
 * dynamic updates
 */

isc_boolean_t
dlz_ssumatch(const char *signer, const char *name, const char *tcpaddr,
             const char *type, const char *key, uint32_t keydatalen,
             uint8_t *keydata, void *dbdata);

/*
 * dlz_addrdataset() is optional, but must be supplied if you want to
 * support dynamic updates
 */
isc_result_t
dlz_addrdataset(const char *name, const char *rdatastr, void *dbdata,
                void *version);

/* dlz_subrdataset() is optional, but must be supplied if you want to
 * support dynamic updates
 */
isc_result_t
dlz_subrdataset(const char *name, const char *rdatastr, void *dbdata,
                void *version);

/*
 * dlz_delrdataset() is optional, but must be supplied if you want to
 * support dynamic updates
 */
isc_result_t
dlz_delrdataset(const char *name, const char *type, void *dbdata,
                void *version);
